File: backend/app/schemas/outfit.py
"""Outfit generation API schemas."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field

from app.schemas.clothes import ClothesResponse


class Occasion(str, Enum):
    casual = "casual"
    party = "party"
    wedding = "wedding"


class Weather(str, Enum):
    hot = "hot"
    cold = "cold"


class OutfitMode(str, Enum):
    free = "free"
    premium = "premium"


class OutfitLanguage(str, Enum):
    """UI / explanation language for matcher reasons and stylist text."""

    en = "en"
    hi = "hi"


class OutfitGenerateRequest(BaseModel):
    """Context for rule-based outfit matching."""

    occasion: Occasion = Field(..., description="Social context")
    weather: Weather = Field(..., description="Temperature context")
    mode: OutfitMode = Field(
        default=OutfitMode.free,
        description="Premium mode: stronger style/color weighting and deterministic picks (requires subscription)",
    )
    language: OutfitLanguage = Field(
        default=OutfitLanguage.en,
        description="Language for matcher reason lines and AI explanations",
    )


class OutfitQuotaInfo(BaseModel):
    """Daily generation usage (free tier has a cap; premium is unlimited)."""

    used_today: int = Field(..., ge=0, description="Sessions counted today after this response")
    daily_limit: int | None = Field(
        None,
        description="Max sessions per day for free users; null when unlimited",
    )


class OutfitUsageResponse(BaseModel):
    """Current user's outfit generation allowance (before starting a new generation)."""

    is_premium: bool
    used_today: int = Field(..., ge=0)
    daily_limit: int | None = Field(None, description="Null when unlimited (premium)")


class OutfitGenerateResponse(BaseModel):
    """One generated outfit (bottom omitted when the upper piece is a dress)."""

    top: ClothesResponse = Field(..., description="Top or dress")
    bottom: ClothesResponse | None = Field(
        default=None,
        description="Bottom piece; null when `top` is a dress",
    )
    footwear: ClothesResponse = Field(..., description="Shoes or boots")
    reasons: list[str] = Field(
        default_factory=list,
        description="Why this outfit works (color, style, weather, occasion)",
    )
    quota: OutfitQuotaInfo = Field(..., description="Remaining daily allowance for free tier")


class OutfitFeedbackRequest(BaseModel):
    """Record feedback for a specific combination of wardrobe item IDs."""

    top_id: int = Field(..., description="Top or dress item id")
    bottom_id: int | None = Field(
        default=None,
        description="Bottom id; omit or null when the top is a dress",
    )
    footwear_id: int = Field(..., description="Footwear item id")
    liked: bool = Field(..., description="True if the user liked this combination")


class OutfitFeedbackResponse(BaseModel):
    """Acknowledgement of stored feedback."""

    model_config = {"from_attributes": True}

    id: int
    liked: bool
    created_at: datetime


class OutfitStylistRequest(BaseModel):
    """Inputs for stylist mode."""

    occasion: Occasion = Field(..., description="Social context")
    vibe: str = Field(..., min_length=1, max_length=50, description="Overall vibe direction (e.g. classy)")
    language: OutfitLanguage = Field(
        default=OutfitLanguage.en,
        description="Language for matcher reason lines (stylist option labels stay stable; localize in UI)",
    )


class StylistOutfit(BaseModel):
    """Outfit triple without feedback metadata."""

    top: ClothesResponse = Field(..., description="Top or dress")
    bottom: ClothesResponse | None = Field(
        default=None,
        description="Bottom piece; null when `top` is a dress",
    )
    footwear: ClothesResponse = Field(..., description="Shoes or boots")


class StylistOptionLabel(str, Enum):
    minimal = "Minimal"
    trendy = "Trendy"
    bold = "Bold"


class StylistOption(BaseModel):
    label: StylistOptionLabel
    outfit: StylistOutfit


class OutfitStylistResponse(BaseModel):
    options: list[StylistOption]
    quota: OutfitQuotaInfo = Field(..., description="Daily generation usage after this response")
    stylist_tier: str = Field(
        ...,
        description="'full' (three options) for Premium; 'preview' (one option) for free",
    )


class OutfitAssistantRequest(BaseModel):
    """Natural-language outfit request (e.g. \"casual college outfit for hot day\")."""

    query: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Describe the outfit you want; occasion, vibe, and weather are inferred",
    )
    mode: OutfitMode = Field(
        default=OutfitMode.free,
        description="Premium: stronger matching when user has subscription",
    )
    language: OutfitLanguage = Field(
        default=OutfitLanguage.en,
        description="Language for stylist explanation and matcher reasons",
    )


class OutfitAssistantResponse(BaseModel):
    """Generated outfit plus a stylist explanation (LLM when configured)."""

    outfit: StylistOutfit
    explanation: str = Field(..., description="Why this outfit fits the request")
    quota: OutfitQuotaInfo = Field(..., description="Daily generation usage after this response")


class SavedOutfitSaveRequest(BaseModel):
    """Bookmark an outfit by wardrobe item ids (same shape rules as feedback)."""

    top_id: int = Field(..., description="Top or dress item id")
    bottom_id: int | None = Field(
        default=None,
        description="Bottom id; omit or null when the top is a dress",
    )
    footwear_id: int = Field(..., description="Footwear item id")


class SavedOutfitItemResponse(BaseModel):
    """One saved outfit with resolved clothing rows."""

    model_config = {"from_attributes": True}

    id: int
    created_at: datetime
    top: ClothesResponse
    bottom: ClothesResponse | None
    footwear: ClothesResponse

File: frontend/src/lib/api/outfit.ts
import { apiClient } from '@/lib/api/client'
import type { ClothesType } from '@/lib/api/types'
import { useUiStore } from '@/stores/useUiStore'

export type AppLanguage = 'en' | 'hi'

function appLanguage(): AppLanguage {
  return useUiStore.getState().language
}

export const OCCASIONS = ['casual', 'party', 'wedding'] as const
export type Occasion = (typeof OCCASIONS)[number]

export const WEATHERS = ['hot', 'cold'] as const
export type Weather = (typeof WEATHERS)[number]

export const STYLIST_VIBES = ['classy', 'trendy', 'bold'] as const
export type StylistVibe = (typeof STYLIST_VIBES)[number]

/** Shape returned by FastAPI `ClothesResponse` (field `clothes_type`). */
export type ClothesApiPiece = {
  id: number
  user_id: number
  image_url: string
  clothes_type: ClothesType
  color: string
  style: string
  created_at: string
}

export type OutfitQuotaApi = {
  used_today: number
  daily_limit: number | null
}

export type OutfitGenerateApiResponse = {
  top: ClothesApiPiece
  bottom: ClothesApiPiece | null
  footwear: ClothesApiPiece
  /** Matcher rationale lines from `POST /outfit/generate` (optional for older responses). */
  reasons?: string[]
  quota?: OutfitQuotaApi
}

export type OutfitStylistApiResponse = {
  options: StylistOptionApiResponse[]
  quota: OutfitQuotaApi
  stylist_tier: 'full' | 'preview'
}

export type OutfitUsageApiResponse = {
  is_premium: boolean
  used_today: number
  daily_limit: number | null
}

export type StylistOptionLabel = 'Minimal' | 'Trendy' | 'Bold'

export type StylistOptionApiResponse = {
  label: StylistOptionLabel
  outfit: OutfitGenerateApiResponse
}

export type OutfitPiece = {
  id: number
  image_url: string
  type: ClothesType
  color: string
  style: string
}

export type GeneratedOutfit = {
  top: OutfitPiece
  bottom: OutfitPiece | null
  footwear: OutfitPiece
}

export type StylistOption = {
  label: StylistOptionLabel
  outfit: GeneratedOutfit
}

/** Mapped outfit + raw API payload (required for `POST /outfit/feedback`). */
export type GenerateOutfitBundle = {
  outfit: GeneratedOutfit
  raw: OutfitGenerateApiResponse
  /** Copy of matcher `reasons` for confidence UI (not sent with feedback). */
  reasons: string[]
  quota?: OutfitQuotaApi
}

/** Map rationale count to a stable match score for display. */
export function matchConfidenceFromReasons(reasons: string[]): number {
  const n = reasons.length
  return Math.min(97, Math.max(72, 72 + Math.min(25, n * 3)))
}

export function stylistOptionConfidence(label: StylistOptionLabel): number {
  switch (label) {
    case 'Minimal':
      return 94
    case 'Trendy':
      return 87
    case 'Bold':
      return 90
    default:
      return 85
  }
}

export type StylistPickLabelKey = 'minimal' | 'trendy' | 'bold'

export function stylistPickLabelKey(label: StylistOptionLabel): StylistPickLabelKey {
  switch (label) {
    case 'Minimal':
      return 'minimal'
    case 'Trendy':
      return 'trendy'
    case 'Bold':
      return 'bold'
    default:
      return 'trendy'
  }
}

function mapPiece(p: ClothesApiPiece): OutfitPiece {
  return {
    id: p.id,
    image_url: p.image_url,
    type: p.clothes_type,
    color: p.color,
    style: p.style,
  }
}

export async function fetchOutfitUsage(): Promise<OutfitUsageApiResponse> {
  const { data } = await apiClient.get<OutfitUsageApiResponse>('/outfit/usage')
  return data
}

export async function generateOutfitRequest(
  occasion: Occasion,
  weather: Weather
): Promise<GenerateOutfitBundle> {
  const { data } = await apiClient.post<OutfitGenerateApiResponse>('/outfit/generate', {
    occasion,
    weather,
    language: appLanguage(),
  })
  return {
    raw: data,
    reasons: data.reasons ?? [],
    quota: data.quota,
    outfit: {
      top: mapPiece(data.top),
      bottom: data.bottom ? mapPiece(data.bottom) : null,
      footwear: mapPiece(data.footwear),
    },
  }
}

export async function generateStylistRequest(
  occasion: Occasion,
  vibe: StylistVibe
): Promise<{ options: StylistOption[]; stylist_tier: 'full' | 'preview'; quota: OutfitQuotaApi }> {
  const { data } = await apiClient.post<OutfitStylistApiResponse>('/outfit/stylist', {
    occasion,
    vibe,
    language: appLanguage(),
  })

  return {
    stylist_tier: data.stylist_tier,
    quota: data.quota,
    options: data.options.map((opt) => ({
      label: opt.label,
      outfit: {
        top: mapPiece(opt.outfit.top),
        bottom: opt.outfit.bottom ? mapPiece(opt.outfit.bottom) : null,
        footwear: mapPiece(opt.outfit.footwear),
      },
    })),
  }
}

export async function submitOutfitFeedbackRequest(
  rawOutfit: OutfitGenerateApiResponse,
  liked: boolean
): Promise<void> {
  await apiClient.post('/outfit/feedback', {
    top_id: rawOutfit.top.id,
    bottom_id: rawOutfit.bottom?.id ?? null,
    footwear_id: rawOutfit.footwear.id,
    liked,
  })
}

export type SavedOutfitApiItem = {
  id: number
  created_at: string
  top: ClothesApiPiece
  bottom: ClothesApiPiece | null
  footwear: ClothesApiPiece
}

export async function saveOutfitRequest(body: {
  top_id: number
  bottom_id: number | null
  footwear_id: number
}): Promise<SavedOutfitApiItem> {
  const { data } = await apiClient.post<SavedOutfitApiItem>('/outfit/save', body)
  return data
}

export async function fetchSavedOutfitsRequest(): Promise<SavedOutfitApiItem[]> {
  const { data } = await apiClient.get<SavedOutfitApiItem[]>('/outfit/saved')
  return data
}

export async function deleteSavedOutfitRequest(savedOutfitId: number): Promise<void> {
  await apiClient.delete(`/outfit/saved/${savedOutfitId}`)
}
